resharper/src/resharper-unity/CSharp/Daemon/Stages/Analysis/UseExplicitTypeInsteadOfStringAnalyzer.cs

```diff
@@ -0,0 +1,175 @@
+        private static readonly IDictionary<string, TypeFilter> InterestingMethods = new Dictionary<string, TypeFilter>()
+        {
+            {"GetComponent", GetComponentTypeFilter},
+            {"AddComponent", AddComponentTypeFilter},
+            {"CreateInstance", CreateInstanceTypeFilter}
+        };
+        
+        private static readonly ISet<IClrTypeName> InterestingClasses = new HashSet<IClrTypeName>()
+        {
+            KnownTypes.ScriptableObject,
+            KnownTypes.GameObject,
+            KnownTypes.Component,
+        };
+        
+        public UseExplicitTypeInsteadStringAnalyzer(UnityApi unityApi)
+            : base(unityApi)
+        {
+        }
+
+        protected override void Analyze(IInvocationExpression expression, ElementProblemAnalyzerData data,
+            IHighlightingConsumer consumer)
+        {
+            if (!expression.IsValid()) return;
+            if (!ValidityChecker.IsValidExpression(expression.GetText())) return;
+            
+            var reference = expression.Reference;
+            if (expression.TypeArguments.Count != 0 || expression.Arguments.Count != 1 || reference == null) return;
+
+            var methodName = reference.GetName();
+            if (!InterestingMethods.ContainsKey(methodName)) return;
+
+            var containingClrTypeName = GetContainingClrTypeName(reference);
+            if (!InterestingClasses.Contains(containingClrTypeName)) return;
+            
+            var argument = expression.Arguments.First().Value;
+            var stringLiteral = GetStringLiteralFromArgument(argument);
+            if (stringLiteral == null) return;
+
+            var types = ResolveStringLiteral(stringLiteral, methodName, argument);
+            var filteredTypes = types.Where(t => InterestingMethods[methodName](t)).ToArray();
+            
+            if (filteredTypes.Length == 0 && types.Any())
+            {
+                var inheritedName = containingClrTypeName.Equals(KnownTypes.ScriptableObject)
+                    ? KnownTypes.ScriptableObject
+                    : KnownTypes.MonoBehaviour;
+                consumer.AddHighlighting(new OnlyInvalidTypeInsteadOfStringWarning(argument as ILiteralExpression, stringLiteral, inheritedName.FullName));
+                return;
+            }
+            
+            if (filteredTypes.Length == 0)
+            {
+                consumer.AddHighlighting(new InvalidStringForTypeWarning(argument as ILiteralExpression, stringLiteral));
+                return;
+            }
+
+            if (filteredTypes.Length >= 2)
+            {
+                consumer.AddHighlighting(new AmbiguousUnityMonoBehaviourDefinitionWarning(argument as ILiteralExpression, stringLiteral));
+            }
+            
+            consumer.AddHighlighting(new UseExplicitTypeInsteadOfStringUsingWarning(expression, methodName, stringLiteral, argument, filteredTypes));
```

Comment:
Please add a check for preprocessor directives.`if (!expression.ContainsPreprocessorDirectives())`
Current implementation might suggest incorrect changes in cases like this one:
```cs
class Foo : MonoBehavior { }
class Test : MonoBehavior
{
  public object M()
  {
    return GetComponent(
#if DEBUG
    "DebugType"
#else
    "ReleaseType"
#endif
    );
  }
}
```

I believe it's worth supporting preprocessor here as such requests for components look like a convenient point to customize debug/release behavior.